### PR TITLE
~~Test~~ Demo using data from WW R package

### PR DESCRIPTION
**Pull request by gvegayon**
This PR does the following:

- [X] It introduces the `datasets` submodule that includes test data generated by the WW team (see [here](https://github.com/CDCgov/wastewater-informed-covid-forecasting/blob/292526383ece582f10823fc939c7e590ca349c6d/cfaforecastrenewalww/data/example_df.rda))
- [X] It includes a test asserting that the wastewater data is properly loaded. It uses polars to load the data.
~~- [ ] It uses `HospitalizationsModel` to fit the data and compares it with the estimates obtained from the R package.~~
- [X] It adds a demo using a polars dataset (with the WW data).
- [X] Adds an example creating a weekday effect.
- [X] Adds a `padding` argument to the hospital admissions and basic renewal model. The padding assumes the first `padding` observations are `nan`.

image

## From the 🏰 Gargoyleosaurus sprint (April 12)

Part of the PR initially wanted to deal with adding a test comparing the results of the WW hospitalization model in R. Still, the plans changed as the number of codebase modifications grew substantially. Things that need to be addressed during the next sprint:

- ~~Check the initialization of the model: The first half of NUTS estimated hospitalizations doesn't cover the data properly:~~

image

- Add Rt with damping effect (see #45).
- Improve the current implementation of the weekday effect: This version of the package has only a toy example (see #14).
- Implement the latent infections with feedback (see #9).

cc @dylanhmorris, @SamuelBrand1

### CODE REVIEW

**Comment by gvegayon** [comment]
I'm holding making progress on this one until #60 is merged.

File: model/docs/example-with-datasets.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+<div><style>
+.dataframe > thead > tr,
+.dataframe > tbody > tr {
+  text-align: right;
+  white-space: pre-wrap;
+}
+</style>
+<small>shape: (5, 3)</small>
+
+| date         | hospitalizations | population |
```
Comment (gvegayon) [inline model/docs/example-with-datasets.md:68]: Oh, there's a typo
```suggestion
| date         | hospitalizations | population
```

File: model/src/pyrenew/datasets/example_df.R
Comment (dylanhmorris) [inline model/src/pyrenew/datasets/example_df.R:1]: Would be nice to do this with pure python. This one will be the most difficult because it reads an `.rda`, but could use something like this? https://rdata.readthedocs.io/en/stable/index.html

File: model/src/pyrenew/datasets/generation_interval.R
Comment (dylanhmorris) [inline model/src/pyrenew/datasets/generation_interval.R:1]: Why not use pure python here?

File: model/src/pyrenew/datasets/infection_admission_interval.R
Comment (dylanhmorris) [inline model/src/pyrenew/datasets/infection_admission_interval.R:1]: Why not pure python here?

File: model/src/pyrenew/observation/negativebinomial.py [hunk 2]
```diff
@@ -32,6 +33,9 @@ def __init__(
             Suffix for the numpy variable.
         parameter_name : str, optional
             Name for the numpy variable.
+        eps : float, optional
+            Small value to add to the predicted mean to prevent numerical
+            instability.
```
Comment (dylanhmorris) [inline model/src/pyrenew/observation/negativebinomial.py:38]: nice idea

File: model/src/pyrenew/datasets/example_df.R
Comment (gvegayon) [reply model/src/pyrenew/datasets/example_df.R:1]: Agree, but that code is just as a reference to show how we got the data. Since it's not a python file and it is included in the poetry toml, I don't think it gets imported with pyrenew. That said, I can add a note on that and maybe save that information in a subfolder to make things cleaner.

File: model/src/pyrenew/datasets/infection_admission_interval.R
Comment (gvegayon) [reply model/src/pyrenew/datasets/infection_admission_interval.R:1]: Implemented in e749c74

File: model/src/pyrenew/datasets/generation_interval.R
Comment (gvegayon) [reply model/src/pyrenew/datasets/generation_interval.R:1]: Implemented in e749c74

File: model/src/pyrenew/datasets/data-raw/generation_interval.R
Comment (dylanhmorris) [inline model/src/pyrenew/datasets/data-raw/generation_interval.R:1]: this can now be deleted

File: model/src/pyrenew/datasets/data-raw/generation_interval.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Dataset from:
+# https://raw.githubusercontent.com/CDCgov/wastewater-informed-covid-forecasting/0962c5d1652787479ac72caebf076ab55fe4e10c/input/saved_pmfs/generation_interval.csv
+import polars as pl
+
+gen_int = pl.read_csv(
+    "https://raw.githubusercontent.com/CDCgov/wastewater-informed-covid-forecasting/0962c5d1652787479ac72caebf076ab55fe4e10c/input/saved_pmfs/generation_interval.csv",
+)
+
+gen_int.write_csv(
+    file="src/pyrenew/datasets/gen_int.tsv",
```
Comment (dylanhmorris) [inline model/src/pyrenew/datasets/data-raw/generation_interval.py:10]: use `os.path.join` here

Comment (dylanhmorris) [reply model/src/pyrenew/datasets/data-raw/generation_interval.py:10]: also ensure directory exists?

File: model/src/pyrenew/datasets/data-raw/infection_admission_interval.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+# https://raw.githubusercontent.com/CDCgov/wastewater-informed-covid-forecasting/0962c5d1652787479ac72caebf076ab55fe4e10c/input/saved_pmfs/inf_to_hosp.csv
+import polars as pl
+
+# Read CSV file from URL
+url = "https://raw.githubusercontent.com/CDCgov/wastewater-informed-covid-forecasting/0962c5d1652787479ac72caebf076ab55fe4e10c/input/saved_pmfs/inf_to_hosp.csv"
+infection_admission_interval = pl.read_csv(url)
+
+# Save as TSV
+infection_admission_interval.write_csv(
+    file="src/pyrenew/datasets/infection_admission_interval.tsv",
```
Comment (dylanhmorris) [inline model/src/pyrenew/datasets/data-raw/infection_admission_interval.py:11]: `os.path.join` and ensure directory?

Comment (gvegayon) [reply model/src/pyrenew/datasets/data-raw/infection_admission_interval.py]: Addressed in 338dee0

File: model/src/pyrenew/datasets/data-raw/generation_interval.py (hunk 3 above)
Comment (gvegayon) [reply model/src/pyrenew/datasets/data-raw/generation_interval.py]: Addressed in 338dee0

File: model/src/pyrenew/datasets/data-raw/generation_interval.R
Comment (gvegayon) [reply model/src/pyrenew/datasets/data-raw/generation_interval.R:1]: Addressed in 338dee0

**Comment by gvegayon** [comment]
@dylanhmorris and @AFg6K7h4fhy2, this PR is ready for review. The only problem is a test that checks the `Model.plot_posterior` function is able to reproduce an image. I had the good intention to check reproducibility (relevant to https://github.com/CDCgov/multisignal-epi-inference/issues/81#issuecomment-2057729045,) but the test fails. I'm not sure why. Any ideas?

**Comment by SamuelBrand1** [comment]
Some context: @gvegayon and I just re-ran with the 
```math
\log \mathcal{R}_t - \log \mathcal{R}_{t-1} \sim N(0., \sigma)
```
$\sigma$ value set at $\sigma = 0.1$ (large!) as opposed to $ \sigma = 0.025$, and that repairs a chunk of the posterior predictive fit; as expected by @dylanhmorris .

Obviously, that can't fix the whole problem because of how the infection -> hosp obs. delay works.

**Comment by gvegayon** [comment]
OK, @dylanhmorris, this is ready for review.

File: model/src/pyrenew/model/hospitalizations.py [hunk 5]
```diff
@@ -87,7 +88,7 @@ def __init__(
             gen_int=gen_int,
             I0=I0,
             latent_infections=latent_infections,
-            observed_infections=DeterministicVariable((0,)),
+            observed_infections=DeterministicVariable(0),
```
[the remark this answers is not in the document]
Comment (gvegayon) [reply model/src/pyrenew/model/hospitalizations.py:91]: Yes, I got confused by the name myself! I agree that none should be a better option. I'm pushing an update to reflect that in a few minutes. The thing about the NullObservationProcess is that it would return None from the sample. I think that's the cleanest.

[the remark this answers is not in the document]
Comment (gvegayon) [reply model/src/pyrenew/model/hospitalizations.py]: Partly addressed now, opened an issue to deal with more general cases across the package #85.

File: model/src/pyrenew/model/rtinfectionsrenewal.py [hunk 6]
```diff
@@ -292,12 +301,26 @@ def sample(
...
         # Using the predicted infections to sample from the observation process
-        observed, *_ = self.sample_infections_obs(
-            predicted=latent, observed_infections=observed_infections, **kwargs
-        )
+        if (observed_infections is not None) and (padding > 0):
+            sampled_pad = jnp.repeat(jnp.nan, padding)
+
+            sampled_obs, *_ = self.sample_infections_obs(
+                predicted=latent[padding:],
+                observed_infections=observed_infections[padding],
```
[the remark this answers is not in the document]
Comment (gvegayon) [reply model/src/pyrenew/model/rtinfectionsrenewal.py]: Fixed!

File: model/docs/example-with-datasets.md [hunk 7]
```diff
@@ -323,7 +329,7 @@ hosp_model.run(
     observed_hospitalizations=dat2,
     rng_key=jax.random.PRNGKey(54),
     mcmc_args=dict(progress_bar=False),
-    padding=days_to_inpute,
+    padding=days_to_inpute, # Padding the model
```
Comment (seabbs) [inline model/docs/example-with-datasets.md:332]: Impute?

File: model/src/pyrenew/datasets/data-raw/generation_interval.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+# Building path to save the file
+path = os.path.join(
+    "src",
+    "pyrenew",
+    "datasets",
+    "gen_int.tsv",
+)
+
+assert os.path.exists(os.path.dirname(path))
```
Comment (dylanhmorris) [inline model/src/pyrenew/datasets/data-raw/generation_interval.py:19]: Should create if not exist. Can use this, which is equivalent to `mkdir -p` at the command line
```suggestion
os.makedirs(os.path.dirname(path), exist_ok=True)
```

File: model/src/pyrenew/datasets/data-raw/infection_admission_interval.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+# Building path to save the file
+path = os.path.join(
+    "src",
+    "pyrenew",
+    "datasets",
+    "infection_admission_interval.tsv",
+)
+
+assert os.path.exists(os.path.dirname(path))
```
Comment (dylanhmorris) [inline model/src/pyrenew/datasets/data-raw/infection_admission_interval.py:19]: Should create if not exist. Can use this, which is equivalent to `mkdir -p` at the command line
```suggestion
os.makedirs(os.path.dirname(path), exist_ok=True)
```

File: model/src/pyrenew/datasets/data-raw/wastewater.R [hunk 10]
```diff
@@ -0,0 +1,13 @@
+load(
+  "https://github.com/CDCgov/wastewater-informed-covid-forecasting/blob/292526383ece582f10823fc939c7e590ca349c6d/cfaforecastrenewalww/data/example_df.rda"
+  )
+
+# Saving as TSV
+write.table(
+  x = example_df,
+  file = "src/pyrenew/datasets/wastewater.tsv",
```
Comment (dylanhmorris) [inline model/src/pyrenew/datasets/data-raw/wastewater.R:8]:
```suggestion
  file = file.path(datasets_dir "wastewater.tsv"),
```

File: model/src/pyrenew/datasets/data-raw/wastewater.R [hunk 11]
```diff
@@ -0,0 +1,13 @@
+load(
+  "https://github.com/CDCgov/wastewater-informed-covid-forecasting/blob/292526383ece582f10823fc939c7e590ca349c6d/cfaforecastrenewalww/data/example_df.rda"
+  )
+
```
Comment (dylanhmorris) [inline model/src/pyrenew/datasets/data-raw/wastewater.R:4]:
```suggestion
datasets_dir <- file.path("src", "pyrenew", "datasets")
dir.create(datasets_dir)

```

File: model/src/pyrenew/model/hospitalizations.py [hunk 12]
```diff
@@ -87,27 +87,27 @@ def __init__(
...
         )
 
         HospitalizationsModel.validate(
-            latent_hospitalizations, observed_hospitalizations
+            latent_hospitalizations, observation_process
         )
 
         self.latent_hospitalizations = latent_hospitalizations
-        self.observed_hospitalizations = observed_hospitalizations
+        self.observation_process = observation_process
```
Comment (dylanhmorris) [inline model/src/pyrenew/model/hospitalizations.py:99]: Don't we need to overwrite with a `NullObservationProcess` or similar if set to `None`?

File: model/src/pyrenew/model/rtinfectionsrenewal.py [hunk 13]
```diff
@@ -71,22 +72,22 @@ def __init__(
...
-            observed_infections=observed_infections,
+            observation_process=observation_process,
             Rt_process=Rt_process,
         )
 
         self.gen_int = gen_int
         self.i0 = I0
         self.latent_infections = latent_infections
-        self.observed_infections = observed_infections
+        self.observation_process = observation_process
```
Comment (dylanhmorris) [inline model/src/pyrenew/model/rtinfectionsrenewal.py:82]: Again, should be overwritten with a null process if None is passed, yes?

File: model/src/pyrenew/model/rtinfectionsrenewal.py [hunk 14]
```diff
@@ -292,12 +297,29 @@ def sample(
         )
 
         # Using the predicted infections to sample from the observation process
-        observed, *_ = self.sample_infections_obs(
-            predicted=latent, observed_infections=observed_infections, **kwargs
-        )
+        if self.observation_process is not None:
```
Comment (dylanhmorris) [inline model/src/pyrenew/model/rtinfectionsrenewal.py:300]: Oh I see you're using an if/else for now. I think it's better to take the `NullProcess` approach in the long run, but this is okay for now. I didn't see the equivalent for the admissions model though, so my comment there still applies.

**Review by dylanhmorris — COMMENTED** [review]
A few key questions, mostly looks ready @gvegayon

File: model/docs/example-with-datasets.md (hunk 7 above)
Comment (gvegayon) [reply model/docs/example-with-datasets.md]: Addressed in 4fc97ed

File: model/docs/example-with-datasets.md [hunk 15]
```diff
@@ -199,6 +199,10 @@ latent_hosp = latent.HospitalAdmissions(
     )
 ```
 
+    /mnt/c/Users/xrd4/Documents/repos/msr/model/.venv/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
+      from .autonotebook import tqdm as notebook_tqdm
+    An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.
```
Comment (dylanhmorris) [inline model/docs/example-with-datasets.md:204]: Can suppress this warning with `numpyro.set_platform("cpu")`

Comment (gvegayon) [reply model/docs/example-with-datasets.md:204]: Since I got your approval, I'll merge this and create an issue to add that line to the examples.

File: model/src/pyrenew/model/hospitalizations.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gvegayon) [reply model/src/pyrenew/model/hospitalizations.py]: PR #93 covers this now.
